### ISSUE TITLE
Package request: pgvecto.rs

### ISSUE THREAD

**Issue by jcollie** [NixOS/nixpkgs#274509]
**Project description**

Scalable Vector database plugin for Postgres, written in Rust, specifically designed for LLM

Immich 1.91 requires this extension for it's PostreSQL database.

**Metadata**

* homepage URL: https://github.com/tensorchord/pgvecto.rs
* source URL: https://github.com/tensorchord/pgvecto.rs
* license: asl20
* platforms: unix, linux, darwin, ...

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by anoadragon453** [comment]
While it doesn't build from source, I've come across this derivation of `pgvecto.rs` which packages the appropriate binary release: https://github.com/diogotcorreia/dotfiles/blob/7676201683a3785ef17eff9f4ad3295375c670bd/packages/pgvecto-rs.nix

**Comment by jcollie** [comment]
@anoadragon453 thanks for that link! That should get me going for now until a native package can be added to nixpkgs.

**Comment by alexmoras** [comment]
Did anyone get this working? I'm quite new to Nix and haven't got a clue where to start with installing a custom `.nix` package.

**Comment by anoadragon453** [comment]
Hi @alexmoras. Yes, I've got it working [in my config](https://github.com/anoadragon453/dotfiles/commit/de37bcd64f702b16115fb405f559e979a1e0260e#diff-8f732ff9aa6533343d2b0f42228f4091570a67c74472a2dce97786f89407d07d), and presumably @diogotcorreia has as well (whose dotfiles repo I linked to).

Typically to include a custom nix package in your config, you merge it into the upstream nixpkgs that your config is already pulling from by using [overlays](https://nixos.wiki/wiki/Overlays). I do this in my config - my [pkgs/default.nix](https://github.com/anoadragon453/dotfiles/blob/de37bcd64f702b16115fb405f559e979a1e0260e/pkgs/default.nix#L70-L128) file is just a big function with `self: super:` arguments (aka an overlay). I then [import that file](https://github.com/anoadragon453/dotfiles/blob/de37bcd64f702b16115fb405f559e979a1e0260e/flake.nix#L42-L58) (so `localpkgs` is now an overlay) and then pass that in the `overlays` list when instantiating nixpkgs. (`lib.mkPkgs` is [yet another custom function](https://github.com/anoadragon453/dotfiles/blob/de37bcd64f702b16115fb405f559e979a1e0260e/lib/default.nix#L68-L73)).

@diogotcorreia does things a bit differently. They have [a package definition](https://github.com/diogotcorreia/dotfiles/blob/7676201683a3785ef17eff9f4ad3295375c670bd/packages/pgvecto-rs.nix) which is a function with arguments `{ lib, stdenv, fetchurl, dpkg, postgresql }:`, which returns a derivation (created via [stdenv.mkDerivation](https://ryantm.github.io/nixpkgs/stdenv/stdenv)).

This then has [`callPackage` called on it, before wrapping it into an overlay](https://github.com/diogotcorreia/dotfiles/blob/7676201683a3785ef17eff9f4ad3295375c670bd/overlays/packages.nix#L9-L15) before that overlay is finally [used when creating a nixpkgs intstance](https://github.com/diogotcorreia/dotfiles/blob/7676201683a3785ef17eff9f4ad3295375c670bd/flake.nix#L86-L97).

So in short, you have a package derivation (either the output of `stdenv.mkDerivation` or `nixpkgs.callPackage` (or other functions)). You need to get it into an overlay function, then pass that overlay function when creating your instance of nixpkgs.

I appreciate this is not simple :) Ideally someone can PR this to nixpkgs to make this easier for everyone.

I learned myself by piecing together bits of other user's configs. If you want more fine-grained help, check out [the NixOS matrix chatroom](https://matrix.to/#/#nix:nixos.org).

**Comment by diogotcorreia** [comment]
If no one else picks this up in the next few days I might give it a shot if I have time, but as I said in my comment https://github.com/NixOS/nixpkgs/pull/244803#issuecomment-1858306506 this would need `cargo-pgrx` to be packaged as well.

Thanks @anoadragon453 for the amazing explanation above!

**Comment by jcollie** [comment]
@diogotcorreia [cargo-pgrx](pkgs/development/tools/rust/cargo-pgrx/default.nix) is already packaged. One problem that I ran into is that cargo-pgrx is built with rust-stable while pgvecto.rs seems to need to be built with rust-nightly.

@alexmoras you can see how I got pgvecto.rs working in my [flake](https://git.ocjtech.us/jeff/nixos-postgresql).

**Comment by diogotcorreia** [comment]
Hey!

I've just managed to package `pgvecto.rs` version `0.1.14-beta` (and it took me way too long)! Unfortunately, I was getting a weird linker error on `0.1.13`, so I went straight for the beta instead.
I'll be keeping an eye for when it releases as a stable version (i.e., `0.1.14`).
I have to say that I have not tested it in Immich yet, but it builds and enables correctly on PostgreSQL.

With that said, I'll be submitting a draft PR shortly, so that I can start getting reviews.

Thank you @jcollie for telling me about `buildPgrxExtension`, that helped a lot. I got around the rust nightly limitation by using the `RUSTC_BOOTSTRAP` env variable and a few patches.

**Comment by diogotcorreia** [comment]
@EricTheMagician That's not good news :sob: 
I do have my repo at https://github.com/diogotcorreia/dotfiles, but I have not pushed yet.

This issue seems to indicate that this will be temporary and they plan to upgrade to a newer version when it's more stable: https://github.com/immich-app/immich/issues/5911

**Comment by diogotcorreia** [comment]
Hey!
I've (almost) finished the PR to add pgvecto.rs 0.2.0 (which will be the targeted version in the next version of Immich, see https://github.com/NixOS/nixpkgs/pull/281192#issuecomment-1937355090).

I have left some comments about stuff I'm not certain about, so if you have the knowledge to do so, I'd appreciate if you could go there and leave a review! https://github.com/NixOS/nixpkgs/pull/281192

Thank you!
